docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,207 @@
+might be handled in a "standard" way (middleware) alongside the others.
+
+We propose middleware-solution, which could help developers implement the following cases:
+* add external logging (let's say sending reports to external services like [Sentry](https://sentry.io));
+* call panic for specific error cases;
+
+It will also make `OutOfGas` case and `default` case one of the middlewares.
+`Default` case wraps recovery object to an error and logs it ([example middleware implementation](#Recovery-middleware)).
+
+## Decision
+
+### Design
+
+#### Overview
+
+Instead of hardcoding custom error handling into BaseApp we suggest using set of middlewares which can be customized
+externally and will allow developers use as many custom error handlers as they want. Implementation with tests
+can be found [here](https://github.com/cosmos/cosmos-sdk/pull/6053).
+
+#### Implementation details
+
+##### Recovery handler
+
+We add a `recover()` object handler type:
+
+```go
+type RecoveryHandler func(recoveryObj interface{}) error
+```
+
+`recoveryObj` is a return value of `recover()` function.
+Handler should type assert (or other methods) an object to define if object should be handled.
+`nil` should be returned if input object can't be handled by that `RecoveryHandler` (not a handler's target type).
+Not `nil` error should be returned if input object was handled and middleware chain execution should be stopped.
+
+An example:
+
+```go
+func exampleErrHandler(recoveryObj interface{}) error {
+    err, ok := recoveryObj.(error)
+    if !ok { return nil }
+    
+    if someSpecificError.Is(err) {
+        panic(customPanicMsg)
+    } else {
+        return nil
+    }
+}
+```
+
+This example breaks the application execution, but it also might enrich the error's context like the `OutOfGas` handler.
+
+##### Recovery middleware
+
+We also add a middleware type (decorator). That function type wraps `RecoveryHandler` and returns next middleware in
+execution chain and handler's `error`. Type is used to separate actual `recovery()` object handling from middleware
+chain processing.
+
+```go
+type recoveryMiddleware func(recoveryObj interface{}) (recoveryMiddleware, error)
+
+func newRecoveryMiddleware(handler RecoveryHandler, next recoveryMiddleware) recoveryMiddleware {
+    return func(recoveryObj interface{}) (recoveryMiddleware, error) {
+        if err := handler(recoveryObj); err != nil {
+            return nil, err
+        }
+        return next, nil
+    }
+}
+```
+
+Function receives a `recover()` object and returns:
+* (next `recoveryMiddleware`, `nil`) if object wasn't handled (not a target type) by `RecoveryHandler`;
+* (`nil`, not nil `error`) if input object was handled and other middlewares in the chain should not be executed;
+* (`nil`, `nil`) this is an invalid behaviour 'cause in that case panic recovery might not be properly handled;
```

Comment:
This can happen when there is not handler that accepts/knows about a given `recoveryObj`, right? We should prohibit this. We can, for instance, always append/ensure a correct "terminator" handler exists at the end. Kind of like a fail-safe.

Reply:
Yes you are right. On `BaseApp` creation (default constructor) default middleware is always created, so this situation won't happen actually. I'll make a note about this.

Reply:
Actually that has been already mentioned in the "BaseApp changes" section.